stream_alert/rules_engine/rules_engine.py

```diff
@@ -211,6 +207,119 @@ def _rule_analysis(self, payload, rule):
 
         return alert
 
+    def _configure_outputs(self, rule):
+        # Check if the rule is staged and, if so, only use the required alert outputs
+        if rule.is_staged(self._rule_table):
+            all_outputs = self._required_outputs_set
+        else:  # Otherwise, combine the required alert outputs with the ones for this rule
+            all_outputs = self._required_outputs_set.union(rule.outputs_set)
+
+        return all_outputs
+
+    @staticmethod
+    def _configure_publishers(rule):
+        """Assigns publishers to each output.
+
+        The @Rule publisher syntax accepts several formats, including a more permissive blanket
+        option.
+
+        In this configuration we DELIBERATELY do not include required_outputs as required outputs
+        should never have their alerts transformed.
+
+        Args:
+            rule (Rule): The rule to create publishers for
+
+        Returns:
+            dict: Maps string outputs names to lists of strings of their fully qualified publishers
+        """
+        def standardize_publisher_list(list_of_references):
+            """Standardizes a list of requested publishers"""
+            publisher_names = map(standardize_publisher_name, list_of_references)
+
+            # Filter out None from the array
+            return [x for x in publisher_names if x is not None]
+
+        def standardize_publisher_name(string_or_reference):
+            """Standardizes a requested publisher into a string name
+
+            Requested publishers can be either the fully qualified string name, OR it can be a
+            direct reference to the function or class.
+            """
+            if not is_publisher_declaration(string_or_reference):
+                LOGGER.error('Invalid publisher requested: %s', string_or_reference)
+                return None
+
+            if isinstance(string_or_reference, basestring):
```

Comment:
Maybe I misunderstood during this whole review and the intention is to support both string literal references and objects.

Reply:
Maybe I should only support references? The string literals are a relic of a bygone age at this point

Reply:
To be clear, there's actually 2 places where code lives to support "publishers"

* On the @rule(publishers=...), which permissively supports both string + references, but I'm trending toward just references
* On any code that pulls from `alert.publishers`, which sits in *DynamoDb*; this code cannot support references at all and can only support strings

Reply:
++ for just references to reduce complexity and chance for fat-fingering typed literals